habitat_code/ramachandran_code_for_review.py

```diff
@@ -0,0 +1,511 @@
+          "which tracts have been downloaded and processed...")
+
+    # Extract tracts to be calculated
+    all_tracts = tracts_w_census_gdf['TRACTCE'].astype('int64')
+
+    # Load tracts already processed
+    all_greenspace_stats_df = pd.read_csv(all_greenspace_stats_path)
+
+    # Check which values in tract_numbers are not downloaded
+    missing_tracts = all_tracts[~all_tracts.isin(all_greenspace_stats_df['tract'])]
+
+    # Create a DataFrame with the missing tracts and their geometries
+    tract_geometry = tracts_w_census_gdf[['TRACTCE', 'geometry']]
+    tract_geometry['TRACTCE'] = tract_geometry['TRACTCE'].astype('int64')
+
+    missing_tracts_df = pd.DataFrame({'missing_tract': missing_tracts.astype('int64')})
+
+    missing_tracts_merged_df = missing_tracts_df.merge(
+        tract_geometry,
+        left_on='missing_tract',
+        right_on='TRACTCE',
+        how='left'
+    )
+
+    missing_tracts_gdf = gpd.GeoDataFrame(missing_tracts_merged_df, geometry='geometry')
+
+    missing_tracts_gdf.head()
+
+    # Plot unprocessed tracts
+    ax = missing_tracts_gdf.plot(color='purple', label='Unprocessed tracts')
+
+    # Adding title
+    plt.title('Unprocessed tracts')
+
+    plt.show()
+
+else:
+    all_greenspace_stats_df = pd.DataFrame(columns=['tract', 'fraction_greenspace'])
+    all_greenspace_stats_df.to_csv(all_greenspace_stats_path, index=False)
+
+    print("All greenspace stats file does not exist. Created file.")
+    
+
+
+# %%
+# Compile image URLs for all tracts that are yet to be processed
+
+naip_image_urls_path = os.path.join(data_dir, 'naip_image_urls.csv')
+year = 2021
+item_url_dfs = []
+
+# Search for image for each tract
+for index, tract in missing_tracts_gdf.iterrows():
+    # print(tract) 
+
+    tract_name = tract['TRACTCE']
+
+    print(tract_name)
+
+    # Search catalog for image
+    tract_geometry = tract['geometry']
+
+    naip_search = pc_catalog.search(
+    collections=["naip"],
+    intersects=shapely.to_geojson(tract_geometry),
+    datetime=f"{year}"
+    )
+
+    print(naip_search)
+
+    try:
+
+        for naip_item in naip_search.items():
+
+            print(naip_item.id)
+            item_url_dfs.append(
+                pd.DataFrame(dict(
+                    tract=[tract_name],
+                    tile_id=[naip_item.id],
+                    url=naip_item.assets['image'].href
+                ))
+            )
+        
+    except Exception as e:
+        print(f"Error processing item: {str(e)}")
+            
+        continue
+
+
+item_url_df = pd.concat(item_url_dfs)
+item_url_df
+
+# Save URLS
+item_url_df.to_csv(naip_image_urls_path, index=False)
+
+# %%
+# Download and process data for all tracts
+
+ndvi_threshold = 0.12
+
+print("All greenspace stats file does not exist. Running computation.")
+
+all_greenspace_stats = []
+
+# Calculate greenspace fraction for each tract
+for tract, tract_urls in item_url_df.groupby('tract'):
```

Comment:
This loop is written very clearly -- I like having the comments and spacing in here. It is, however, a tad long for a single loop, so you might consider splitting it up into a few functions.